misc/tetramesh/src/processors/tetrameshvolumeraycaster.cpp

```diff
@@ -104,28 +104,36 @@ void TetraMeshVolumeRaycaster::initializeResources() {
 }
 
 void TetraMeshVolumeRaycaster::process() {
-    TetraMeshProvider* meshProvider{*inport_.getData()};
-    if (!meshProvider) {
-        throw Exception(IVW_CONTEXT, "Tetra mesh does not have a valid data provider");
+    if (inport_.isChanged() || !mesh_) {
+        const auto& tetraMesh = *inport_.getData();
+
+        auto nodes = tetraMesh.getNodes();
+        auto nodeIds = tetraMesh.getNodeIds();
+        utiltetra::fixFaceOrientation(nodes, nodeIds);
+
+        auto opposingFaces = utiltetra::getOpposingFaces(nodeIds);
+
+        buffers_.upload(nodes, nodeIds, opposingFaces);
```

Comment:
pass in the tertaMesh?

Reply:
That causes some stuff to be evaluated twice. `nodes` and `nodeIds` are generated on the fly by the interface (no references). The mesh also needs opposing faces.

Reply:
Seems like this code is in several places?

Reply:
Which was one reason for holding the buffer and boundary mesh in the data instead of the vectors.